### PR TITLE
feat: resource tree filters

### PR DESCRIPTION
**Pull request by arunjaindev**
# Description
fix: retain filter on resource tab change
fix: redirect to all if a filter is no more available
fix: switch to available kind when filter is switched
use TabGroup component in node details tabs
use common button component in app details source info
fix: redirection to source pipeline in case of unauthorized

Fixes https://github.com/devtron-labs/sprint-tasks/issues/1041

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/v2/appDetails/k8Resource/nodeType/NodeTree.component.tsx [hunk 1]
```diff
@@ -40,26 +40,31 @@ const NodeTreeComponent = ({
...
         }
-        const _clickedNodes = generateSelectedNodes(clickedNodes, _treeNodes, _node, parents, isDevtronApp)
+        const _clickedNodes = generateSelectedNodes(clickedNodes, treeNodes, _node, parents, isDevtronApp)
         registerNodeClick(_clickedNodes)
         setReRender(!reRender)
     }
 
     useEffect(() => {
-        const _urlArray = window.location.href.split(`${URLS.APP_DETAILS_K8}/`)
+        const _urlArray = window.location.pathname.split(`${URLS.APP_DETAILS_K8}/`)
```
Comment (AbhishekA1509) [inline src/components/v2/appDetails/k8Resource/nodeType/NodeTree.component.tsx:67]: Can we use location from useLocation instead?

File: src/components/v2/appDetails/k8Resource/K8Resource.component.tsx [hunk 2]
```diff
@@ -41,12 +48,89 @@ export const K8ResourceComponent = ({
...
+    // nodes according to current filter
+    const currentFilteredNodes = useMemo(
+        () =>
+            nodes.filter(
+                (node) =>
+                    currentFilter === ALL_RESOURCE_KIND_FILTER ||
+                    node.health?.status.toLowerCase() === currentFilter ||
+                    (currentFilter === NodeFilters.drifted && node.hasDrift),
+            ),
+        [currentFilter],
```
Comment (AbhishekA1509) [inline src/components/v2/appDetails/k8Resource/K8Resource.component.tsx:70]: Do we need nodes in dep?

File: src/components/v2/appDetails/k8Resource/K8Resource.component.tsx [hunk 3]
```diff
@@ -41,12 +48,89 @@ export const K8ResourceComponent = ({
...
+        }
+    }, [nodes])
+
+    const getPodNameForSelectedFilter = (selectedFilter: string) => {
+        const podParents = getPodsRootParentNameAndStatus(nodes)
+        const selectNode = podParents.find((parent) => parent[1].toLowerCase() === selectedFilter)?.[0]
+        return selectNode?.split('/')?.[2]
+    }
+
+    const appDetails = IndexStore.getAppDetails()
```
Comment (AbhishekA1509) [inline src/components/v2/appDetails/k8Resource/K8Resource.component.tsx:89]: This is only being used in getRedirectPathname, can move it inside handler

File: src/components/v2/appDetails/k8Resource/K8Resource.component.tsx [hunk 4]
```diff
@@ -41,12 +48,89 @@ export const K8ResourceComponent = ({
...
+
+    const handleFilterClick = (selectedFilter: string) => {
+        const searchParams = new URLSearchParams([['filterType', selectedFilter]])
+        if (selectedFilter === ALL_RESOURCE_KIND_FILTER) {
+            history.push({ search: `${searchParams}` })
+            return
+        }
+        // current selected node exist in new selected filter or not
+        const nextFilterNodes = nodes.filter(
+            (node) =>
```
Comment (AbhishekA1509) [inline src/components/v2/appDetails/k8Resource/K8Resource.component.tsx:117]: These checks seems duplicated from useMemo, can move to util

File: src/components/v2/appDetails/k8Resource/NodeTreeTabList.tsx [hunk 5]
```diff
@@ -24,12 +24,13 @@ import { ReactComponent as K8ResourceIcon } from '../../../../assets/icons/ic-ob
...
 import { ReactComponent as Cross } from '../../../../assets/icons/ic-close.svg'
 import Tippy from '@tippyjs/react'
-import { ConditionalWrap } from '@devtron-labs/devtron-fe-common-lib'
+import { ConditionalWrap, useSearchString } from '@devtron-labs/devtron-fe-common-lib'
 import './NodeTreeTabList.scss'
 
 export default function NodeTreeTabList({ logSearchTerms, setLogSearchTerms, tabRef }: NodeTreeTabListProps) {
     const { nodeType } = useParams<{ nodeType: string }>()
     const { push } = useHistory()
+    const filterType = useSearchString().searchParams.filterType || 'all'
```
Comment (AbhishekA1509) [inline src/components/v2/appDetails/k8Resource/NodeTreeTabList.tsx:33]: Please use enum/const

File: src/components/v2/appDetails/k8Resource/K8Resource.component.tsx [hunk 6]
```diff
@@ -80,16 +77,18 @@ export const K8ResourceComponent = ({
         }
     }, [nodes])
 
+    useEffect(() => {
```
Comment (AbhishekA1509) [inline src/components/v2/appDetails/k8Resource/K8Resource.component.tsx:80]: Can we handle this without useEffect and can we minimize the dep array?

**Review by AbhishekA1509 — APPROVED** [review]

File: src/components/v2/appDetails/k8Resource/utils.tsx [hunk 7]
```diff
@@ -0,0 +1,4 @@
+import { Node, NodeFilters } from '@devtron-labs/devtron-fe-common-lib'
+
+export const doesNodeSatisfiesFilter = (node: Node, filter: string) =>
+    node.health?.status.toLowerCase() === filter || (filter === NodeFilters.drifted && node.hasDrift)
```
Comment (eshankvaish) [inline src/components/v2/appDetails/k8Resource/utils.tsx:4]: confirm if this shouldn't be in FE lib?

File: src/components/v2/appDetails/k8Resource/K8Resource.component.tsx [hunk 8]
```diff
@@ -41,12 +48,85 @@ export const K8ResourceComponent = ({
...
+    }, [currentFilter, nodes])
+
+    // nodes according to current filter
+    const currentFilteredNodes = useMemo(
+        () =>
+            nodes.filter(
+                (node) => currentFilter === ALL_RESOURCE_KIND_FILTER || doesNodeSatisfiesFilter(node, currentFilter),
+            ),
+        [currentFilter, nodes],
+    )
```
Comment (eshankvaish) [inline src/components/v2/appDetails/k8Resource/K8Resource.component.tsx:72]: verify if this re-computed on each render. I noticed we added `${nodes}` somewhere

**Review by eshankvaish — DISMISSED** [review]

File: src/components/v2/appDetails/k8Resource/utils.tsx (hunk 7 above)
Comment (arunjaindev) [reply src/components/v2/appDetails/k8Resource/utils.tsx:4]: we need to check for filters in oss as well

File: src/components/v2/appDetails/k8Resource/K8Resource.component.tsx (hunk 8 above)
Comment (arunjaindev) [reply src/components/v2/appDetails/k8Resource/K8Resource.component.tsx:72]: not computed on every render, we are getting nodes here from IndexStore

**Review by vivek-devtron — APPROVED** [review]

**Review by eshankvaish — APPROVED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**